src/python_testing/TC_CNET_4_23.py

```diff
@@ -0,0 +1,220 @@
+        return ["CNET.S.F00", "CNET.S.F01"]
+
+    # Override default timeout.
+    @property
+    def default_timeout(self) -> int:
+        return 200
+
+    async def SendConnectNetworkWithFailure(
+            self, networkID: int, endpoint: int):
+        with asserts.assert_raises(ChipStackError) as cm:
+            cmd = Clusters.NetworkCommissioning.Commands.ConnectNetwork(networkID=networkID)
+            await self.send_single_cmd(endpoint=endpoint, cmd=cmd)
+        asserts.assert_equal(cm.exception.err, kChipErrorTimeout, "Unexpected error while trying to send ConnectNetwork command")
+        logging.info("ConnectNetwork command timed out, not assert here as it is expected during network switching")
+
+    async def SendCommissioningCompleteWithRetry(self):
+        commissioning_complete_cmd = Clusters.GeneralCommissioning.Commands.CommissioningComplete()
+        # Implement a retry mechanism for timeout failure.
+        for i in range(kMaxCommissioningCompleteRetryTimes):
+            try:
+                commissioning_complete_results = await self.send_single_cmd(cmd=commissioning_complete_cmd)
+                asserts.assert_true(commissioning_complete_results.errorCode ==
+                                    Clusters.GeneralCommissioning.Enums.CommissioningErrorEnum.kOk, "CommissioningComplete command failed")
+                break
+            except ChipStackError as e:  # chipstack-ok: After switching networks, it may resolve to an incorrect address
+                asserts.assert_equal(e.err, kChipErrorTimeout, "Unexpected error while trying to send CommissioningComplete")
+                logging.info(f"Will retry CommissioningComplete command in 10 seconds, attempt: {i + 1}")
+                time.sleep(10)
+        else:
+            asserts.assert_true(False, "CommissioningComplete command failed")
+
+    @async_test_body
+    async def test_TC_CNET_4_23(self):
+        asserts.assert_true('PIXIT.CNET.THREAD_1ST_OPERATIONALDATASET' in self.matter_test_config.global_test_params,
+                            "PIXIT.CNET.THREAD_1ST_OPERATIONALDATASET must be included on the command line in "
+                            "the --hex-arg flag as PIXIT.CNET.THREAD_1ST_OPERATIONALDATASET:<thread dataset in hex>")
+        self.step("precondition")
+        wifi_ssid = self.get_wifi_ssid().encode("utf-8")
+        wifi_credential = self.get_credentials().encode("utf-8")
+
+        feature_map_response = await self.default_controller.ReadAttribute(self.dut_node_id, [(Clusters.NetworkCommissioning.Attributes.FeatureMap)], fabricFiltered=True)
+        wifi_endpoint = kInvalidEndpointId
+        thread_endpoint = kInvalidEndpointId
+        for endpoint, value in feature_map_response.items():
+            feature_dict = value.get(Clusters.Objects.NetworkCommissioning, {})
+            feature_map = feature_dict.get(Clusters.Objects.NetworkCommissioning.Attributes.FeatureMap, None)
+            if feature_map == kWiFiFeature:
```

Comment:
be careful here - the PDC features are going to land and an equivalence test will break
```suggestion
            if bool(feature_map & Clusters.NetworkCommissioning.Bitmaps.Feature.kWiFiNetworkInterface):
```